### PR TITLE
[SPARK][K8S] Support auto build Kubernetes client from env when Kyuubi running in Pod

### PR DESCRIPTION
**Pull request by zwangsheng**
### _Why are the changes needed?_

Current kyuubi build kubernetes client for `KyuubiApplicationOperator` to find or kill spark engine on kubernetes.

This pr aims to extract the code for building kubernetes client into `KubernetesUtils` and to do auto build following the logic in Spark.

We prefer to building with user-configured parameters; When missing, the kubernetes client logic will build one from environment variables and kubeconfig.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala [hunk 1]
```diff
@@ -135,6 +128,7 @@ class KubernetesApplicationOperation extends ApplicationOperation with Logging {
 
 object KubernetesApplicationOperation extends Logging {
   val LABEL_KYUUBI_UNIQUE_KEY = "kyuubi-unique-tag"
+  val SPARK_APP_SELECTOR = "spark-app-selector"
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala:131]: please change the var name ti align w/ Spark `org.apache.spark.deploy.k8s.Constants`
```
val SPARK_APP_ID_LABEL = "spark-app-selector"
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala [hunk 2]
```diff
@@ -33,24 +34,16 @@ class KubernetesApplicationOperation extends ApplicationOperation with Logging {
...
-        val client = new DefaultKubernetesClient(Config.autoConfigure(contextOpt.get))
-        info(s"Initialized Kubernetes Client connect to: ${client.getMasterUrl}")
-        client
-      } catch {
-        case e: KubernetesClientException =>
-          error("Fail to init KubernetesClient for KubernetesApplicationOperation", e)
-          null
-      }
+
+    info("Start Initialize Kubernetes Client.")
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/KubernetesApplicationOperation.scala:40]:
```suggestion
    info("Start initializing Kubernetes Client.")
```

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 3]
```diff
@@ -828,6 +828,69 @@ object KyuubiConf {
...
+
+  val KUBERNETES_AUTHENTICATE_CLIENT_CERT_FILE: OptionalConfigEntry[String] =
+    buildConf("kyuubi.kubernetes.authenticate.clientCertFile")
+      .doc("kubernetes client authenticate client cert file")
+      .version("1.7.0")
+      .stringConf
+      .createOptional
+
+  val KUBERNETES_AUTHENTICATE_CA_CERT_FILE: OptionalConfigEntry[String] =
+    buildConf("kyuubi.kubernetes.authenticate.ca cert file")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:881]: fix

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 4]
```diff
@@ -828,6 +828,69 @@ object KyuubiConf {
...
+    buildConf("kyuubi.kubernetes.context")
+      .doc("The desired context from your kubernetes config file used to configure the K8S " +
+        "client for interacting with the cluster.")
+      .version("1.6.0")
+      .stringConf
+      .createOptional
+
+  val KUBERNETES_NAMESPACE: ConfigEntry[String] =
+    buildConf("kyuubi.kubernetes.namespace")
+      .version("1.7.0")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:841]: doc

File: kyuubi-server/src/main/scala/org/apache/kyuubi/util/KubernetesUtils.scala [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+import com.google.common.base.Charsets
+import com.google.common.io.Files
+import io.fabric8.kubernetes.client.{Config, ConfigBuilder, DefaultKubernetesClient, KubernetesClient}
+import io.fabric8.kubernetes.client.Config.autoConfigure
+import io.fabric8.kubernetes.client.okhttp.OkHttpClientFactory
+import okhttp3.{Dispatcher, OkHttpClient}
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.config.KyuubiConf.{KUBERNETES_AUTHENTICATE_CA_CERT_FILE, KUBERNETES_AUTHENTICATE_CLIENT_CERT_FILE, KUBERNETES_AUTHENTICATE_CLIENT_KEY_FILE, KUBERNETES_AUTHENTICATE_OAUTH_TOKEN, KUBERNETES_AUTHENTICATE_OAUTH_TOKEN_FILE, KUBERNETES_CONTEXT, KUBERNETES_MASTER, KUBERNETES_NAMESPACE, KUBERNETES_TRUST_CERTIFICATES}
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/util/KubernetesUtils.scala:32]:
```suggestion
import org.apache.kyuubi.config.KyuubiConf._
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/util/KubernetesUtils.scala [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+object KubernetesUtils extends Logging {
+
+  def buildKubernetesClient(conf: KyuubiConf): KubernetesClient = {
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/util/KubernetesUtils.scala:36]: how about returning Option instead?

File: docker/helm/templates/kyuubi-configmap.yaml [hunk 7]
```diff
@@ -47,5 +47,6 @@ data:
     #
     kyuubi.frontend.bind.host={{ .Values.server.bind.host }}
     kyuubi.frontend.bind.port={{ .Values.server.bind.port }}
+    kyuubi.kubernetes.namespace= {{ .Release.Namespace }}
 
     # Details in https://kyuubi.apache.org/docs/latest/deployment/settings.html
```
Comment (pan3793) [inline docker/helm/templates/kyuubi-configmap.yaml:52]: nit: eol

File: kyuubi-server/src/main/scala/org/apache/kyuubi/util/KubernetesUtils.scala [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+      }.withOption(clientKeyFile) {
+        (file, configBuilder) => configBuilder.withClientKeyFile(file)
+      }.withOption(clientCertFile) {
+        (file, configBuilder) => configBuilder.withClientCertFile(file)
+      }.build()
+
+    (master, kubeContext, loadMasterAddressFromEnv) match {
+      case (None, None, Some(url)) =>
+        debug(s"Set kubernetes master address $url from env KUBERNETES_PORT")
+        config.setMasterUrl(url)
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/util/KubernetesUtils.scala:89]: you missed other cases

**Review by pan3793 — COMMENTED** [review]
Overall LGTM, just minor comments

**Comment by yaooqinn** [comment]
I try my best to understand what you are trying to say about this PR in the PR desc, but failed

**Comment by zwangsheng** [comment]
> I try my best to understand what you are trying to say about this PR in the PR desc, but failed

This pr aims to add util object `KubernetesUtils` help kyuubi build kubernetes client in spark way.

**Review by pan3793 — APPROVED** [review]

**Comment by yaooqinn** [comment]
refactor only change?

**Comment by pan3793** [comment]
The PR description is updated and should cover the changes in the PR.

Thanks, merging to master

**Comment by pan3793** [comment]
The CI failure caused by known flaky tests

**Pull request closed without merging**